### PR TITLE
[plugin-db] Add step to wait until data received from SQL request is equal to data from table

### PR DESCRIPTION
**Pull request by tshkursk**
Step wording:
When I wait for '$duration' duration retrying $retryTimes times while
data from `$sqlQuery` executed against `$dbKey` matching rows using
`$keys` is equal to data from:$table

### CODE REVIEW

File: vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java [hunk 1]
```diff
@@ -208,6 +209,56 @@ public void compareData(String sourceSqlQuery, String sourceDbKey, String target
...
+     * </ul>
+     * @param duration Time duration to wait
+     * @param retryTimes How many times request will be retried; duration/retryTimes=timeout between requests
+     * @param sqlQuery SQL query to execute
+     * @param dbKey Key identifying the database connection
+     * @param keys Comma-separated list of column's names to map resulting tables rows
+     * @param table Rows to compare data against
+     */
+    @When("I wait for '$duration' duration retrying $retryTimes times while data from `$sqlQuery`"
+            + " executed against `$dbKey` matching rows using `$keys` is equal to data from:$table")
```
Comment (valfirst) [inline vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java]: do you really need
```
matching rows using `$keys`
```
in this step?

File: vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java [hunk 2]
```diff
@@ -208,6 +209,56 @@ public void compareData(String sourceSqlQuery, String sourceDbKey, String target
...
+                    .map(m -> m.entrySet()
+                               .stream()
+                               .collect(Collectors.toMap(Map.Entry::getKey, e -> String.valueOf(e.getValue())))));
+            result = compareData(statistics, sourceData, targetData);
+            if (result.isEmpty())
+            {
+                break;
+            }
+            Sleeper.sleep(pollingTimeoutMillis, TimeUnit.MILLISECONDS);
+        }
```
Comment (valfirst) [inline vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java]: how does this step look like in the report? isn't it long-running step with single attachment?

File: vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java [hunk 3]
```diff
@@ -208,6 +209,56 @@ public void compareData(String sourceSqlQuery, String sourceDbKey, String target
...
+     *     <li>repeats previous actions if data was not equal and seconds timeout not expired</li>
+     * </ul>
+     * @param duration Time duration to wait
+     * @param retryTimes How many times request will be retried; duration/retryTimes=timeout between requests
+     * @param sqlQuery SQL query to execute
+     * @param dbKey Key identifying the database connection
+     * @param keys Comma-separated list of column's names to map resulting tables rows
+     * @param table Rows to compare data against
+     */
+    @When("I wait for '$duration' duration retrying $retryTimes times while data from `$sqlQuery`"
```
[the remark this answers is not in the document]
Comment (tshkursk) [reply vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java]: Could be wrapped with composite?

File: vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java [hunk 4]
```diff
@@ -208,6 +209,56 @@ public void compareData(String sourceSqlQuery, String sourceDbKey, String target
...
+                               .stream()
+                               .collect(Collectors.toMap(Map.Entry::getKey, e -> String.valueOf(e.getValue())))));
+            result = compareData(statistics, sourceData, targetData);
+            if (result.isEmpty())
+            {
+                break;
+            }
+            Sleeper.sleep(pollingTimeoutMillis, TimeUnit.MILLISECONDS);
+        }
+        while (System.currentTimeMillis() <= expectedTime);
```
Comment (ikalinin1) [inline vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java]: Let's introduce some common class that will handle this do-while waiter logic. We have the same login in Json waiting steps

File: vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (valfirst) [reply vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java]: the problem here is that we can't use nested ExamplesTable-s

File: vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java (hunk 4 above)
Comment (tshkursk) [reply vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java]: Please, specify which class and where you propose?

File: vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java (hunk 1 above)
Comment (tshkursk) [reply vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java]: I thought it's a common matching parameter for comparison steps. Could I remove it?

File: vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java (hunk 2 above)
Comment (tshkursk) [reply vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java]: Yes. I thought no need to attach reports each iteration

File: vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java (hunk 1 above)
Comment (valfirst) [reply vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java]: not really common, if you don't rely on this parameter, you can remove it

File: vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java (hunk 2 above)
Comment (valfirst) [reply vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java]: reports are not needed, that's true, but how can user see that some actions are performed?

File: vividus-plugin-db/src/test/java/org/vividus/bdd/steps/db/DatabaseStepsTests.java [hunk 5]
```diff
@@ -81,6 +81,10 @@
 @SuppressFBWarnings("OBL_UNSATISFIED_OBLIGATION")
 class DatabaseStepsTests
 {
+    private static final String MINUTE_DURATION = "PT1M";
```
Comment (valfirst) [inline vividus-plugin-db/src/test/java/org/vividus/bdd/steps/db/DatabaseStepsTests.java]:
```java
private static final Duration ONE_MINUTE = Duration.ofMinutes(1);
```

Comment (valfirst) [inline vividus-plugin-db/src/test/java/org/vividus/bdd/steps/db/DatabaseStepsTests.java]: 1 minute is too much for unit test

File: vividus-plugin-rest-api/src/main/java/org/vividus/http/HttpRequestData.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.http;
+
+import java.util.Optional;
+
+public class HttpRequestData
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/http/HttpRequestData.java]: class looks generic, but it doesn't support any requests with body

File: vividus-plugin-rest-api/src/main/java/org/vividus/http/HttpRequestExecutor.java [hunk 7]
```diff
@@ -53,52 +55,53 @@ public HttpRequestExecutor(IHttpClient httpClient, HttpTestContext httpTestConte
...
-            }
-            while (repeatRequest.test(httpResponse));
+            Supplier<HttpResponse> responseProvider = () -> {
+                try
+                {
+                    return executeHttpCallSafely(requestData);
+                }
+                catch (IOException e)
+                {
+                    throw new HttpRequestException("An I/O exception occurred", e);
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/http/HttpRequestExecutor.java]: what's the reason to wrap exception? it makes their handling non-transparent

File: vividus-util/src/main/java/org/vividus/util/WaitUtils.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+package org.vividus.util;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+public final class WaitUtils
```
Comment (valfirst) [inline vividus-util/src/main/java/org/vividus/util/WaitUtils.java]: `WaitUtils` -> `Waiter` ?

File: vividus-util/src/main/java/org/vividus/util/WaitUtils.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+{
+    private WaitUtils()
+    {
+        // Nothing to do
+    }
+
+    public static <T> T wait(Duration duration, int retryTimes, Supplier<T> valueProvider,
+            Predicate<T> repeatRequest)
```
Comment (valfirst) [inline vividus-util/src/main/java/org/vividus/util/WaitUtils.java]: `repeatRequest` - name is not clear

File: vividus-plugin-rest-api/src/main/java/org/vividus/http/HttpRequestExecutor.java (hunk 7 above)
Comment (tshkursk) [reply vividus-plugin-rest-api/src/main/java/org/vividus/http/HttpRequestExecutor.java]: Return to a problem with method which throws checked exception and is used in Supplier construction (line 90). Don't know how to avoid now. Custom Supplier looks tricky too.

File: vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ikalinin1) [reply vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java]: @valfirst  what is nested examples table here?

File: vividus-plugin-db/src/test/java/org/vividus/bdd/steps/db/DatabaseStepsTests.java [hunk 10]
```diff
@@ -213,6 +217,7 @@ void shouldExecuteSqlQuery() throws SQLException
         when(dataSources.get(DB_KEY)).thenReturn(dataSource);
         databaseSteps.executeSql(QUERY, DB_KEY);
         assertThat(LOGGER.getLoggingEvents(), equalTo(List.of(info("Executed query: {}\nAffected rows:{}", QUERY, 1))));
+        LOGGER.clear();
```
[the remark this answers is not in the document]
Comment (tshkursk) [reply vividus-plugin-db/src/test/java/org/vividus/bdd/steps/db/DatabaseStepsTests.java]: If not cleared, next test fails

[the remark this answers is not in the document]
Comment (valfirst) [reply vividus-plugin-db/src/test/java/org/vividus/bdd/steps/db/DatabaseStepsTests.java]: - if assertion fails, the logger won't be clean up
 - test execution order is not guaranteed

please investigate the issue and fix the root cause

[the remark this answers is not in the document]
Comment (ikalinin1) [reply vividus-plugin-db/src/test/java/org/vividus/bdd/steps/db/DatabaseStepsTests.java]: TestLoggerExtension should  help ?

File: vividus-plugin-rest-api/src/main/java/org/vividus/http/HttpRequestExecutor.java (hunk 7 above)
Comment (ikalinin1) [reply vividus-plugin-rest-api/src/main/java/org/vividus/http/HttpRequestExecutor.java]: @valfirst  Can we start to store them into some utility package? :)

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java [hunk 11]
```diff
@@ -215,7 +214,7 @@ public void saveElementsNumberByJsonPath(String jsonPath, Set<VariableScope> sco
      * <ul>
      * <li>Executes HTTP GET request to the specified resource</li>
      * <li>Checks if response body contains an element by JSON path</li>
-     * <li>Sleeps during one tenth part of specified duration</li>
+     * <li>Sleeps during during calculated part of specified duration</li>
```
Comment (ikalinin1) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: shouldn't it go as separate commit?

Comment (tshkursk) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: I have updates on this step. So I guess javadoc in scope.

File: vividus-plugin-rest-api/src/main/java/org/vividus/http/HttpRequestParameters.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.http;
+
+import java.util.Optional;
+
+public class HttpRequestParameters
```
[the remark this answers is not in the document]
Comment (tshkursk) [reply vividus-plugin-rest-api/src/main/java/org/vividus/http/HttpRequestParameters.java]: Removed

File: vividus-util/src/main/java/org/vividus/util/wait/Waiter.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.function.Predicate;
+
+public final class Waiter
+{
+    private Waiter()
+    {
+        // Nothing to do
+    }
+
+    public static <T, E extends Exception> T wait(Duration duration, int retryTimes,
```
[the remark this answers is not in the document]
Comment (tshkursk) [reply vividus-util/src/main/java/org/vividus/util/wait/Waiter.java]: Added

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java (hunk 11 above)
Comment (ikalinin1) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java]: But this change has no relation to the step update :)

File: vividus-util/src/main/java/org/vividus/util/wait/WaitMode.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+ * limitations under the License.
+ */
+
+package org.vividus.util.wait;
+
+import java.time.Duration;
+
+public class WaitMode
+{
+    private Duration duration;
```
Comment (valfirst) [inline vividus-util/src/main/java/org/vividus/util/wait/WaitMode.java]: `final`

File: vividus-util/src/main/java/org/vividus/util/wait/WaitMode.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    private int retryTimes;
```
Comment (valfirst) [inline vividus-util/src/main/java/org/vividus/util/wait/WaitMode.java]: `final`

File: vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java [hunk 16]
```diff
@@ -208,6 +212,54 @@ public void compareData(String sourceSqlQuery, String sourceDbKey, String target
...
+        statistics.getTarget().setRowsQuantity(sourceData.size());
+
+        CheckedSupplier<List<List<EntryComparisonResult>>, Exception> resultProvider = () -> {
+            List<Map<String, Object>> data = jdbcTemplate.queryForList(sqlQuery);
+            statistics.getSource().setRowsQuantity(data.size());
+            Map<Object, Map<String, Object>> targetData = hashMap(Set.of(), data.stream()
+                    .map(m -> m.entrySet()
+                               .stream()
+                               .collect(Collectors.toMap(Map.Entry::getKey, e -> String.valueOf(e.getValue())))));
+            return compareData(statistics, sourceData, targetData);
```
Comment (uarlouski) [inline vividus-plugin-db/src/main/java/org/vividus/bdd/steps/db/DatabaseSteps.java]: you may use Collectors.collectingAndThen without caching targetData
